docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,76 @@
+
+## Proposed Solution
+
+We will introduce a new object, the Connection Manager, which will be responsible for managing all probes. These probes are stored as distinct states within an enumeration class, `ConnectionManager::State`. The Connection Manager will sequentially cycle through each of these states in a systematic order, only proceeding to the next state if the current probe is successful.
+
+In the event of a probe failure, the Connection Manager will emit a signal to the frontend. This will trigger a modal to be displayed to the user, providing an explanation of the issue and suggesting potential solutions. Upon encountering a failed probe, the progression through the remaining probes is immediately halted. Once the issue has been resolved, the cycle of probes commences anew. If there are no ongoing probes, the Connection Manager will enter an Idle state.
+
+The Connection Health, an existing component of our codebase, is tasked with sending periodic pings and determining the stability of the VPN based on the responses received (see `ConnectionHealth::healthCheckup()`). It classifies the VPN stability as `Stable`, `Unstable` (some missing responses), or `NoSignal` (no responses received). When the VPN experiences issues with network connectivity or server availability, it naturally transitions into the No Signal state and, ideally, recovers once the issue is resolved. We can leverage this logic to recover from a probe failure within the Connection Manager.
+
+## Implementation Plan
+
+```c++
+enum ConnectionManagerStates {
+  InternetProbe,
+  Firewall,
+  ServerProbe,
+  CheckSubscription,
+  Idle,
+}
+```
+### Upon VPN Activation (`StateOff` to `StateConnecting`)
+Perform a single cycle going through all the Connection Manager probes.
+- If a probe fails:
+  - Set Controller to `StateOff`
+  - Stop subsequent probes from running
+  - emit a signal to the frontend communicating which probe failed
+- If all probes pass:
+  - Activate VPN (Controller `StateOn`)
+
+### VPN is already active (`StateOn`)
+Perform periodic checks cycling through all the Connection Manager probes.
+- If a probe fails:
+  - Controller continues to stay in `StateOn`. We will __never__ toggle the VPN off without the user explicitly doing so themselves
+  - We will __not__ manipulate the Connection Health stability based on Connection Manager probes
+  - Stop subsequent probes from running
+  - emit a signal to the frontend communicating which probe failed
+- If all probes pass:
+  - Do nothing. VPN stays on and stable
+
+> Note: The second scenario where a check fails when the VPN is already active still needs some more thinking. There will likely be a follow up design doc on how we will approach that later on.
```

Comment:
What do you mean by "needs some more thinking"? Can you add example of what downsides we might be pushing to a future doc?

Reply:
Not any downsides necessary, we just need to work out the specific details that I don't want to get into in this document.

Reply:
To me, this document is where we should be thinking this through, to make sure we don't back ourselves into a corner. This seems really closely tied to the decisions being made in this doc, so why shouldn't it be further explored here?